### DISCUSSION TITLE
Symfony & Dependency injection

### DISCUSSION THREAD

**Discussion by PaulCombal** [nutgram/nutgram#666] [Q&A] [answered]
Hello,

I am using a middleware (`SaveTelegramUserMiddleware`) that uses dependency injection to access a Repository:

// App/Telegram/Middleware/SaveTelegramUserMiddleware
```php
<?php

namespace App\Telegram\Middleware;

use App\Entity\TelegramUser;
use App\Repository\TelegramUserRepository;
use SergiX44\Nutgram\Nutgram;

class SaveTelegramUserMiddleware
{
    protected TelegramUserRepository $telegramUserRepository;

    public function __construct(TelegramUserRepository $telegramUserRepository)
    {
        $this->telegramUserRepository = $telegramUserRepository;
    }

    public function __invoke(Nutgram $bot, $next): void
    {
        // ...
        $this->telegramUserRepository->updateOrCreateUser($user);
        $next($bot);
    }
}
```

When I'm running my bot, which uses this middleware, in `telegram.php` using `bin/console nutgram:run`, everything works wonderfully.

However, I am creating a bot object somewhere else in the code, like so:
```php
        $bot = new Nutgram($token, $configuration);
        // ...
        $bot->middleware(SaveTelegramUserMiddleware::class);
```

And every time I try using that bot using `$bot->run()`, an exception gets thrown as soon as the middleware is invoked:
`SergiX44\Container\Exception\ContainerException: Cannot resolve parameter '$registry::App\Repository\TelegramUserRepository' in ContainerException.php:18`

My question is: How do you use dependency injection properly with Nutgram? How can I use the service container with programmatically created bots?

And most importantly: is this documented?

Thank you!

PS: Symfony 7.0.3

**Comment by Lukasss93** [comment, marked as answer]
Hi, yes, there is an undocumented configuration: "container".
This configuration is done automatically on Laravel and Symfony frameworks.
If you want to use an external service container, you have to set it in Nutgram's configuration or use its container.

### Using Nutgram's container:
```php
$bot = new Nutgram('your-token');
$bot->getContainer()->set(CustomService::class, new CustomService());
``` 

### Set a custom container:
```php
$config = new Configuration(container: $yourContainer);
$bot = new Nutgram('your-token', $config);
```
This mode is done automatically by the frameworks:
- https://github.com/nutgram/symfony-bundle/blob/master/src/DependencyInjection/Factory/NutgramFactory.php#L36
- https://github.com/nutgram/laravel/blob/master/src/NutgramServiceProvider.php#L38
